cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSQLEngineAdapter.java

```diff
@@ -132,30 +236,207 @@ public BatchSQLEngineAdapter(SQLEngine<?, ?, ?, ?> sqlEngine) {
+   *
+   * @param joinDefinition the join definition to check
+   * @return Collection containing all the {@link SQLDataset} instances needed to execute this join.
+   */
+  @SuppressWarnings("unchecked")
+  private Collection<SQLDataset> getJoinInputDatasets(JoinDefinition joinDefinition) throws IllegalArgumentException {
+    List<SQLDataset> datasets = new ArrayList<>(joinDefinition.getStages().size());
+
+    for (JoinStage stage : joinDefinition.getStages()) {
+      // Wait for the previous push or execute jobs to complete
+      SQLEngineJobKey pushJobKey = new SQLEngineJobKey(stage.getStageName(), SQLEngineJobType.PUSH);
+      SQLEngineJobKey execJobKey = new SQLEngineJobKey(stage.getStageName(), SQLEngineJobType.EXECUTE);
+
+      if (jobs.containsKey(pushJobKey)) {
+        SQLEngineJob<SQLDataset> job = (SQLEngineJob<SQLDataset>) jobs.get(pushJobKey);
+        waitForJobAndHandleExceptionInternal(job);
+        datasets.add(job.waitFor());
+      } else if (jobs.containsKey(execJobKey)) {
+        SQLEngineJob<SQLDataset> job = (SQLEngineJob<SQLDataset>) jobs.get(execJobKey);
+        waitForJobAndHandleExceptionInternal(job);
+        datasets.add(job.waitFor());
+      } else {
+        throw new IllegalArgumentException("No SQL Engine job exists for stage " + stage.getStageName());
+      }
+    }
+
+    return datasets;
+  }
+
+  /**
+   * Join implementation. This method has blocking calls and should be executed in a separate thread.
+   *
+   * @param future the future instance to use to return results.
+   * @param joinRequest the Join Request
+   * @throws SQLEngineException   if any of the preceding jobs fails.
+   */
+  private void joinInternal(CompletableFuture<SQLDataset> future,
+                            SQLJoinRequest joinRequest)
+    throws SQLEngineException {
+
+    String datasetName = joinRequest.getDatasetName();
+    DefaultStageMetrics stageMetrics = new DefaultStageMetrics(metrics, datasetName);
+    StageStatisticsCollector statisticsCollector = statsCollectors.get(datasetName);
+
+    // Count input metrics for each of the preceding stages.
+    for (SQLDataset inputDataset : joinRequest.getInputDatasets()) {
+      countRecordsIn(inputDataset, statisticsCollector, stageMetrics);
+    }
+
+    //Execute Join job.
+    SQLDataset joinDataset = (SQLDataset) sqlEngine.join(joinRequest);
+
+    // Count output rows and complete future.
+    countRecordsOut(joinDataset, statisticsCollector, stageMetrics);
+    future.complete(joinDataset);
+  }
+
   /**
    * Stops all jobs from executing and cleans up the SQL engine.
    */
-  public void cancel() {
-    for (String dataset : jobs.keySet()) {
-      jobs.get(dataset).cancel();
-      sqlEngine.cleanup(dataset);
+  public void stop() throws SQLEngineException {
+    Set<String> datasetNames = new HashSet<>();
+    // Stop all jobs
+    for (SQLEngineJob<?> job : jobs.values()) {
+      job.cancel();
```

Comment:
can this throw an exception? If so, would be good to catch and still try to cancel everything else before throwing an exception. If multiple throw exception, should use addSuppressed on the exception getting thrown.

Reply:
Good point. Added logic to capture all exceptions (if any),

Reply:
Added logic to catch any and all exceptions thrown by this method.

I also followed Vitalii's suggestion and implemented the Closeable interface in this class and renamed `stop` to `close`.